zcash_client_sqlite/src/wallet.rs

```diff
@@ -2363,43 +2363,94 @@ pub(crate) fn set_transaction_status(
     conn: &rusqlite::Transaction,
     params: &P,
-    block_height: BlockHeight,
-) -> Result<(), SqliteClientError> {
-    let sapling_activation_height = params
-        .activation_height(NetworkUpgrade::Sapling)
-        .expect("Sapling activation height must be available.");
+    requested_height: BlockHeight,
+) -> Result<BlockHeight, SqliteClientError> {
+    // Determine a checkpoint to which we can rewind, if any.
+    #[cfg(not(feature = "orchard"))]
+    let truncation_checkpoint = conn
+        .query_row(
+            "SELECT MAX(checkpoint_id) FROM sapling_tree_checkpoints
+             WHERE checkpoint_id <= :block_height",
+            named_params! {":block_height": u32::from(requested_height)},
+            |row| {
+                row.get::<_, Option<u32>>(0)
+                    .map(|opt| opt.map(BlockHeight::from))
+            },
+        )
+        .optional()?
+        .flatten();
+
+    #[cfg(feature = "orchard")]
+    let truncation_checkpoint = conn
+        .query_row(
+            "SELECT MAX(stc.checkpoint_id)
+             FROM sapling_tree_checkpoints stc
+             JOIN orchard_tree_checkpoints otc ON stc.checkpoint_id = otc.checkpoint_id
+             WHERE stc.checkpoint_id <= :block_height",
+            named_params! {":block_height": u32::from(requested_height)},
+            |row| {
+                row.get::<_, Option<u32>>(0)
+                    .map(|opt| opt.map(BlockHeight::from))
+            },
+        )
+        .optional()?
+        .flatten();
+
+    let truncation_height = truncation_checkpoint.map_or_else(
+        || {
+            let min_truncation_height = conn
+                .query_row(
+                    "SELECT MIN(checkpoint_id) FROM sapling_tree_checkpoints",
+                    [],
+                    |row| {
+                        row.get::<_, Option<u32>>(0)
+                            .map(|opt| opt.map(BlockHeight::from))
+                    },
+                )
+                .optional()?
+                .flatten();
+
+            Err(SqliteClientError::RequestedRewindInvalid(
+                min_truncation_height,
+                requested_height,
+            ))
+        },
+        Ok,
+    )?;
 
-    // Recall where we synced up to previously.
     let last_scanned_height = conn.query_row("SELECT MAX(height) FROM blocks", [], |row| {
-        row.get::<_, Option<u32>>(0)
-            .map(|opt| opt.map_or_else(|| sapling_activation_height - 1, BlockHeight::from))
+        let h = row.get::<_, Option<u32>>(0)?;
+
+        Ok(h.map_or_else(
+            || {
+                params
+                    .activation_height(NetworkUpgrade::Sapling)
+                    .expect("Sapling activation height must be available.")
+                    - 1
+            },
+            BlockHeight::from,
+        ))
     })?;
 
-    if block_height < last_scanned_height - PRUNING_DEPTH {
-        if let Some(h) = get_min_unspent_height(conn)? {
```

Comment:
It was never the case that we could actually truncate down to no checkpoints - we could only remove the _last_ checkpoint, but leave the tree in the state that it was as of the last checkpoint (which is another foot-gun that is now fixed.)

Reply:
That doesn't match the documentation for `get_min_unspent_height`, which as written conveys the fact that we _specifically_ wanted to handle the case where we didn't have enough checkpoints to rewind to the desired height. The prior intended behaviour was:
```
    |<-- can -->|<-- cannot -->|<---- can rewind ---->|
birthday    min_unspent    earliest_checkpoint     chain_tip
```
where rewinds prior to `min_unspent` were permitted because we never needed to build a witness for any mined note in the wallet as of that height (as they were all spent).

That behaviour was implemented in bb68744df123895272c36107d25127163d18ab1d (#341), and the rename was in ac3439e65ec401b995043c53db99d17594ce99b2 (#827).

If we no longer want that behaviour (or more likely, no longer can _support_ that behaviour, which I suspect might be the case in the presence of non-linear scanning), then at a minimum the now-misleading documentation for `get_min_unspent_height` must be rewritten. And given that the purpose of that method was precisely to enable this use, we should just remove the method entirely if we don't support this use anymore.